src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,212 @@
+
+pub fn handle_direct_volume(vol_cmd: DirectVolumeCommand) -> Result<()> {
+    let command = vol_cmd.directvol_cmd;
+    match command {
+        DirectVolSubcommand::Add(args) => {
+            add(&args.volume_path, &args.mount_info)?;
+        }
+        DirectVolSubcommand::Remove(args) => {
+            remove(&args.volume_path)?;
+        }
+        DirectVolSubcommand::Stats(args) => {
+            let result = executor::block_on(stats(&args.volume_path))?;
+            println!("{}", &result);
+        }
+        DirectVolSubcommand::Resize(args) => {
+            executor::block_on(resize(&args.volume_path, args.resize_size))?;
+        }
+    }
+    Ok(())
+}
+
+async fn resize(volume_path: &str, size: u64) -> Result<()> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volumn_mount_info(volume_path)?;
+    let resize_req = runtimes::ResizeRequest {
+        size,
+        volume_path: mount_info.device,
+    };
+    let encoded = serde_json::to_string(&resize_req)?;
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT)).unwrap();
+
+    let url = runtimes::DIRECT_VOLUMN_RESIZE_URL;
+    let response = shim_client
+        .post(url, &String::from(CONTENT_TYPE_JSON), &encoded)
+        .await?;
+    if response.status() != hyper::StatusCode::OK {
+        return Err(anyhow!(
+            "status {}: failed to resize volume",
+            response.status()
+        ));
+    }
+    Ok(())
+}
+
+async fn stats(volume_path: &str) -> Result<String> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volumn_mount_info(volume_path)?;
+
+    let req_url =
+        url::form_urlencoded::Serializer::new(String::from(runtimes::DIRECT_VOLUMN_STATS_URL))
+            .append_pair(runtimes::DIRECT_VOLUMN_PATH_KEY, &mount_info.device)
+            .finish();
+
+    // sid: sandbox id
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT)).unwrap();
+    shim_client.get(&req_url).await?;
+    let response = shim_client.get(&req_url).await?;
+    // turn body into string
+    let body = hyper::body::to_bytes(response.into_body()).await?;
+    let body = String::from_utf8(body.to_vec()).unwrap();
+    Ok(body)
+}
+
+// join_path joins user provided volumepath with kata direct-volume root path
+fn join_path(volume_path: &str) -> PathBuf {
+    let b64_encoded_path = base64::encode(volume_path.as_bytes());
+    Path::new(KATA_DIRECT_VOLUME_ROOT_PATH).join(b64_encoded_path)
+}
+
+// add writes the mount info (json string) of a direct volume into a filesystem path known to Kata Containers.
+pub fn add(volume_path: &str, mount_info: &str) -> Result<()> {
+    let mount_info_dir_path = join_path(volume_path);
+    // check if path exists and is dir, if not, create this dir
+    if !mount_info_dir_path.exists() {
+        fs::create_dir_all(&mount_info_dir_path)?;
```

Comment:
This is racy. It would be safer to remove the check and call `create_dir_all()` unconditionally.